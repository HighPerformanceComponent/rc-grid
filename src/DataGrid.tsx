import React, {
    cloneElement,
    CSSProperties,
    isValidElement,
    Key,
    ReactNode,
    useContext,
    useMemo,
    useReducer,
    useRef,
    useState,
} from 'react'
import styled from 'styled-components'

import type { Column, DataGridProps, Row } from './types'
import DataGridRow from './Row'
import HeaderRow from './HeaderRow'
import Context, { reducer } from './Context'
import UniversalToolbar from './ plugins/UniversalToolbar'
import { useChevronRightIcon, useChevronDownIcon } from './Icon'

const GridContainer = styled.div`
    position: relative;
`

const Grid = styled.div`
    position: relative;
    overflow: auto;
    border: 1px solid #ddd;
    outline: none;
`

const ExpandableIcon = styled.i`
    width: 16px;
    height: 16px;
    cursor: pointer;
`

function useExpandableRender<T>(
    row: Row<T>,
    isExpandable: (data: Row<T>) => boolean
) {
    const { state, dispatch } = useContext(Context)
    let icon = useChevronRightIcon()
    if (isExpandable?.(row) === false) {
        return null
    }

    const expandable = state.expandableKey.includes(row.key)
    
    if (expandable) {
        icon = useChevronDownIcon()
    }

    return (
        <ExpandableIcon
            onClick={() => {
                const newKeys: Key[] = []
                if (expandable) {
                    state.expandableKey.forEach(ele => {
                        if(ele !== row.key) {
                            newKeys.push(ele)
                        }
                    })
                } else  {
                    newKeys.push(...state.expandableKey, row.key)
                }
                dispatch({
                    type: 'setExpandableKey',
                    payload: newKeys
                })
            }}
        >
            {icon}
        </ExpandableIcon>
    )
}

function DataGrid<R>({
    className,
    style = {},
    rows,
    height = 500,
    width = 1000,
    columns,
    estimatedRowHeight = 50,
    estimatedColumnWidth = 120,
    headerRowHeight = 35,
    cacheRemoveCount = 6,
    defaultColumnWidth = 120,
    expandable,
    onHeaderCellRender,
    onEmptyRowsRenderer,
    onHeaderRowRender = (node: JSX.Element) => node,
    onHeaderResizable,
    onEditorChangeSave,
    onSort,
}: DataGridProps<R>) {
    const [state, dispatch] = useReducer(reducer, {
        editorChange: [],
        sortColumns: [],
        expandableKey: []
    })

    const [universalValue, setUniversalValue] = useState<string>('')
    const [isShowUniversal, setIsShowUniversal] = useState<boolean>(false)

    const gridRef = useRef<HTMLDivElement>(null)

    /** 数据进行排序, 方便固定列进行排序 */
    const sortColumns = useMemo(() => {
        const newColumns = [...columns]
        newColumns.sort((before) => {
            if (before.fixed === 'left') {
                return -1
            }
            if (before.fixed === 'right') {
                return 1
            }
            return 0
        })

        // 如果展开行属性存在, 那么进行新增展开逻辑
        if (expandable) {
            const expandableColumn: Column<R> = {
                name: '$expandable',
                title: '',
                width: 35,
                isSelect: () => false,
                fixed: 'left',
                render: (_text, row) => useExpandableRender(row, expandable?.isExpandable)
            }
            newColumns.splice(0, 0, expandableColumn)
        }
        return newColumns
    }, [columns])

    const filterRows = useMemo(
        () =>
            rows.filter((row) => {
                const { cells } = row
                if (universalValue !== '') {
                    const result = cells.some((cell) => {
                        if (cell.value.indexOf(universalValue) !== -1) {
                            return true
                        }
                        return false
                    })
                    return result
                }
                return true
            }),
        [universalValue, rows]
    )

    // 滚动的高度
    const scrollHeight = useMemo(() => {
        let result = 0
        filterRows.forEach((row) => {
            result += row.height
        })
        return result
    }, [filterRows])

    // 滚动的宽度
    const scrollWidth = useMemo(() => {
        let result = 0
        sortColumns.forEach((column) => {
            result += column.width || defaultColumnWidth
        })
        return result
    }, [sortColumns])

    const startRowTop = useRef<number>(0)
    const [scrollTop, setScrollTop] = useState<number>(0)
    const [scrollLeft, setScrollLeft] = useState<number>(0)
    const [isScroll, setIsScroll] = useState<boolean>(false)

    const calcCacheRemove = estimatedColumnWidth * cacheRemoveCount

    // 渲染表格的行信息
    const renderRow = useMemo(() => {
        // 过滤找到的内容信息
        const domRows: Array<ReactNode> = []
        let top = startRowTop.current
        const headerStyled: CSSProperties = {
            height: headerRowHeight,
            top,
            width: scrollWidth,
        }
        const headerRow: JSX.Element = (
            <HeaderRow
                key="header"
                scrollLeft={scrollLeft}
                scrollWidth={scrollWidth}
                styled={headerStyled}
                gridProps={{
                    className,
                    style,
                    rows,
                    height,
                    width,
                    columns: sortColumns,
                    estimatedRowHeight,
                    estimatedColumnWidth,
                    headerRowHeight,
                    cacheRemoveCount,
                    defaultColumnWidth,
                    onHeaderCellRender,
                    onHeaderResizable,
                    onEmptyRowsRenderer,
                    onHeaderRowRender,
                    onEditorChangeSave,
                    onSort,
                }}
            />
        )

 
        domRows.push(onHeaderRowRender(headerRow))

        top += headerRowHeight
        filterRows.some((row) => {
            const index = rows.findIndex((ele) => ele.key === row.key)
            if (top < scrollTop - calcCacheRemove) {
                top += row.height
                return false
            }
            domRows.push(
                <DataGridRow<R>
                    key={row.key}
                    rows={rows}
                    rowIndex={index}
                    width={width}
                    scrollWidth={scrollWidth}
                    scrollLeft={scrollLeft}
                    styled={{
                        height: row.height,
                        top,
                        width: scrollWidth,
                        lineHeight: `${row.height}px`,
                    }}
                    gridProps={{
                        className,
                        style,
                        rows,
                        height,
                        width,
                        columns: sortColumns,
                        estimatedRowHeight,
                        estimatedColumnWidth,
                        headerRowHeight,
                        cacheRemoveCount,
                        defaultColumnWidth,
                        onHeaderCellRender,
                        onEmptyRowsRenderer,
                        onHeaderRowRender,
                        onHeaderResizable,
                        onEditorChangeSave,
                        onSort,
                    }}
                />
            )

            top += row.height

            // 计算表格的可展开
            if (expandable?.expandedRowRender && state.expandableKey.includes(row.key) ) {
                const expandableElement = expandable?.expandedRowRender(row, {
                    top,
                    width: scrollWidth,
                    position: 'absolute',
                    lineHeight: `${row.height}px`,
                    boxSizing:  'border-box',
                    borderBottom: `1px solid #ddd`,
                })
                if (isValidElement(expandableElement)) {
                    const { style: pStyle = {}, ...restProps} = expandableElement.props
                    const expandableHeight = pStyle.height || 300
                    domRows.push(
                        cloneElement(expandableElement, {
                            key: `${row.key}-expandable`,
                            style: {
                                ...pStyle,
                                height: expandableHeight,
                            },
                            ...restProps
                        })
                    )
                    top += expandableHeight
                }
            }

            if (top > height + scrollTop + calcCacheRemove) {
                return true
            }
            return false
        })
        return domRows
    }, [
        scrollTop,
        scrollLeft,
        filterRows,
        sortColumns,
        estimatedColumnWidth,
        width,
        cacheRemoveCount,
        headerRowHeight,
        rows,
        estimatedRowHeight,
        state.selectPosition,
        state.expandableKey
    ])

    const lastScrollTop = useRef<number>(0)
    const lastScrollLeft = useRef<number>(0)

    const timeout = useRef<ReturnType<typeof setTimeout>>()

    const onScroll = ({
        currentTarget,
    }: React.UIEvent<HTMLDivElement, UIEvent>) => {
        if (timeout.current) {
            clearTimeout(timeout.current)
        } else if (isScroll === false) {
            setIsScroll(true)
        }

        timeout.current = setTimeout(() => {
            setIsScroll(false)
            timeout.current = undefined
        }, 400)

        const { scrollTop: currentScrollTop, scrollLeft: currentScrollLeft } =
            currentTarget
        if (currentTarget) {
            if (
                // 纵向： currentScrollTop - lastScrollTop.current 距离上次滚动的距离
                Math.abs(currentScrollTop - lastScrollTop.current) >
                calcCacheRemove - (estimatedColumnWidth * cacheRemoveCount) / 10
            ) {
                setScrollTop(currentScrollTop)
                lastScrollTop.current = currentTarget.scrollTop
            }

            if (
                // 横向: currentScrollLeft - lastScrollLeft.current 距离上次滚动的距离
                Math.abs(currentScrollLeft - lastScrollLeft.current) >
                calcCacheRemove - (estimatedColumnWidth * cacheRemoveCount) / 10
            ) {
                setScrollLeft(currentScrollLeft)
                lastScrollLeft.current = currentScrollLeft
            }
        }
    }

    const renderUniversal = () => {
        if (isShowUniversal) {
            return (
                <UniversalToolbar
                    value={universalValue}
                    onChange={setUniversalValue}
                    onBlur={() => {
                        gridRef.current.focus()
                        setIsShowUniversal(!isShowUniversal)
                    }}
                />
            )
        }
        return null
    }

    return (
        <Context.Provider
            value={{
                state,
                dispatch,
            }}
        >
            <GridContainer
                style={{
                    width,
                }}
            >
                {renderUniversal()}
                <Grid
                    ref={gridRef}
                    className={className}
                    style={{
                        height,
                        width: '100%',
                        ...style,
                    }}
                    tabIndex={-1}
                    onKeyDown={(e) => {
                        if (e.key === 'f' && e.ctrlKey) {
                            setIsShowUniversal(!isShowUniversal)
                            e.preventDefault()
                        }
                    }}
                    onScroll={onScroll}
                >
                    <div
                        style={{
                            height: scrollHeight,
                            pointerEvents: isScroll ? 'none' : undefined,
                        }}
                    >
                        {renderRow}
                    </div>
                    {rows.length === 0 && onEmptyRowsRenderer ? (
                        <div
                            style={{
                                height: '100%',
                                width,
                                position: 'sticky',
                                left: 0,
                            }}
                        >
                            {onEmptyRowsRenderer()}
                        </div>
                    ) : undefined}
                </Grid>
            </GridContainer>
        </Context.Provider>
    )
}

export default DataGrid
