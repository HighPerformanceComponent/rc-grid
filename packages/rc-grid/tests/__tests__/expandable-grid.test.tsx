import React, { useRef, useState } from 'react'
import { render, waitFor, fireEvent } from "@testing-library/react"
import produce from 'immer'

import DataGrid, { Column, DataGridProps, Row, Cell, EditorProps } from '../../src/index'


const Input = ({ style, value: tempValue, onEditCompleted }: EditorProps) => {
    const [value, setValue] = useState(tempValue)
    return (
        <input
            // eslint-disable-next-line jsx-a11y/no-autofocus
            autoFocus
            style={style}
            value={value as string}
            onBlur={() => {
                onEditCompleted(value)
            }}
            onChange={(e) => {
                const text = e.target.value
                setValue(text)
            }}
        />
    )
}


const rowsData: any[] = []

for (let i = 0; i < 10000; i += 1) {

    const cells: Cell[] = [{
        name: 'id',
        value: `${i}`,
    }, {
        name: 'userName',
        value: `my name ${i}`,
    }, {
        name: 'address',
        value: `address ${i}`,
    }, {
        name: 'email',
        value: `E-Mail ${i}`,
    }, {
        name: 'mobilePhone',
        value: `Mobile Phone ${i}`,
    }]
    const row: Row<any> = {
        height: 35,
        key: `key-${i}`,
        cells,
        object: {
            id: `${i}`,
            userName: `my name ${i}`,
            address: `address ${i}`,
            email: `E-Mail ${i}`,
            mobilePhone: `Mobile Phone ${i}`
        }
    }
    rowsData.push(row)
}


interface BashGridProps<T> extends Omit<DataGridProps<T>, 'rows' | 'columns'> {
    /** 表格的行数据信息 */
    rows?: Row<T>[]
    /** 列的信息 */
    columns?: Column<T>[]
}

function BashGrid<T>(props: BashGridProps<T>) {
    const { columns, rows, ...restProps} = props

    const oldData = useRef<Row<any>[]>(produce(rows, () => { }))
    const [datas, setDatas] = useState<Row<any>[]>(produce(rows, () => { }))
    const [col, setCol] = useState<Column<T>[]>(columns)
    
    return (
        <DataGrid<T>
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...restProps}
            columns={col}
            rows={datas}
            onHeaderResizable={(newCols) => {
                setCol(newCols)
            }}
            expandable={{
                isExpandable: (row) => {
                    const { id } = row.object as any
                    if (id === '1') {
                        return false
                    }
                    return true
                },
                expandedRowRender: (row, style) => (
                    <div style={style}>
                        {' '}
                        这是一个展开的内容信息 {JSON.stringify(row)}{' '}
                    </div>
                ),
            }}
            onSort={(sort) => {
                if (sort.length === 0) {
                    setDatas(oldData.current)
                }
                if (sort.length > 0) {
                    const { direction } = sort[0]
                    const { columnKey } = sort[0]
                    const tempRowData = produce(datas, (newData) => {
                        newData.sort((a, b) => {
                            const aData: string = a.object[columnKey]
                            const bData: string = b.object[columnKey]

                            if (direction === 'ASC') {
                                if (
                                    aData.toUpperCase() >
                                    bData.toUpperCase()
                                ) {
                                    return 1
                                }
                                return -1
                            }
                            if (direction === 'DESC') {
                                if (
                                    aData.toUpperCase() <
                                    bData.toUpperCase()
                                ) {
                                    return 1
                                }
                            }
                            return -1
                        })
                    })
                    setDatas(tempRowData)
                }
            }}
        />
    )
}


BashGrid.defaultProps = {
    columns: [{
        name: 'id',
        title: 'id',
        sort: true,
        resizable: true,
        fixed: 'left'
    }, {
        name: 'userName',
        sort: true,
        resizable: true,
        title: 'User Name',
        editor: Input
    }, {
        name: 'address',
        sort: true,
        resizable: true,
        title: 'Address'
    }, {
        name: 'email',
        sort: true,
        resizable: true,
        title: 'E-Mail'
    }, {
        name: 'mobilePhone',
        title: 'Mobile Phone',
        sort: true,
        resizable: true,
        fixed: 'right'
    }],
    rows: rowsData
}


test('expandable grid test', async () => {
    const { container, getAllByRole } =render((
        <BashGrid />
    ))
    await waitFor(() => getAllByRole('row'))
    const expandableIcon =  getAllByRole('row')[0].children[0].children[0]
    fireEvent.click(expandableIcon)
    fireEvent.click(expandableIcon)
    fireEvent.click(expandableIcon)

    const twoExpandableIcon =  getAllByRole('row')[3].children[0].children[0]
    fireEvent.click(twoExpandableIcon)
    fireEvent.click(twoExpandableIcon)
    fireEvent.click(twoExpandableIcon)
    expect(container).toMatchSnapshot()
})