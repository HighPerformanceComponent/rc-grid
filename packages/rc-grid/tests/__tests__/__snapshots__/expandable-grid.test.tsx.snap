// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandable grid test 1`] = `
<div>
  <div
    class="sc-pNWdM bZoqPE"
    style="width: 1000px; height: 600px;"
  >
    <div
      class="sc-jrsJWt isMknf"
      role="grid"
      style="height: 100%; width: 100%;"
      tabindex="-1"
    >
      <div
        style="height: 350000px;"
      >
        <div
          class="sc-iCoGMd Nmmrq"
          role="rowheader"
          style="height: 35px; top: 0px; width: 635px;"
        >
          <div
            class="sc-eCApnc cibSkH"
            data-id="table-1"
            data-name="$expandable"
            draggable="false"
            role="columnheader"
            style="left: 0px; width: 35px; position: sticky; z-index: 11;"
          >
            <span
              class="sc-gKAaRy"
            />
          </div>
          <div
            class="sc-eCApnc hpuPPE"
            data-id="table-1"
            data-name="id"
            draggable="false"
            role="columnheader"
            style="left: 35px; width: 120px; position: sticky; z-index: 11;"
          >
            <span
              class="sc-gKAaRy"
              style="cursor: pointer;"
            >
              id
            </span>
            <span
              class="sc-jSFjdj jqLned"
            />
          </div>
          <div
            class="sc-eCApnc cibSkH"
            data-id="table-1"
            data-name="userName"
            draggable="false"
            role="columnheader"
            style="left: 155px; width: 120px;"
          >
            <span
              class="sc-gKAaRy"
              style="cursor: pointer;"
            >
              User Name
            </span>
            <span
              class="sc-jSFjdj jqLned"
            />
          </div>
          <div
            class="sc-eCApnc cibSkH"
            data-id="table-1"
            data-name="address"
            draggable="false"
            role="columnheader"
            style="left: 275px; width: 120px;"
          >
            <span
              class="sc-gKAaRy"
              style="cursor: pointer;"
            >
              Address
            </span>
            <span
              class="sc-jSFjdj jqLned"
            />
          </div>
          <div
            class="sc-eCApnc cibSkH"
            data-id="table-1"
            data-name="email"
            draggable="false"
            role="columnheader"
            style="left: 395px; width: 120px;"
          >
            <span
              class="sc-gKAaRy"
              style="cursor: pointer;"
            >
              E-Mail
            </span>
            <span
              class="sc-jSFjdj jqLned"
            />
          </div>
          <div
            class="sc-eCApnc dpazEg"
            data-id="table-1"
            data-name="mobilePhone"
            draggable="false"
            role="columnheader"
            style="width: 120px; position: sticky; z-index: 11; float: right; right: 0px;"
          >
            <span
              class="sc-gKAaRy"
              style="cursor: pointer;"
            >
              Mobile Phone
            </span>
            <span
              class="sc-jSFjdj jqLned"
            />
          </div>
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 35px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            <i
              class="sc-kEqXSa bbXQBg"
            >
              <svg
                class="bi bi-chevron-down"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </i>
          </div>
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            0
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 0
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 0
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 0
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 0
          </div>
        </div>
        <div
          style="top: 70px; width: 635px; position: absolute; line-height: 35px; box-sizing: border-box; border-bottom: 1px solid #ddd; padding: 10px; height: 300px;"
        >
           
          这是一个展开的内容信息 
          {"height":35,"key":"key-0","cells":[{"name":"id","value":"0"},{"name":"userName","value":"my name 0"},{"name":"address","value":"address 0"},{"name":"email","value":"E-Mail 0"},{"name":"mobilePhone","value":"Mobile Phone 0"}],"object":{"id":"0","userName":"my name 0","address":"address 0","email":"E-Mail 0","mobilePhone":"Mobile Phone 0"}}
           
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 370px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          />
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            1
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 1
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 1
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 1
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 1
          </div>
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 405px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            <i
              class="sc-kEqXSa bbXQBg"
            >
              <svg
                class="bi bi-chevron-right"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </i>
          </div>
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            2
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 2
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 2
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 2
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 2
          </div>
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 440px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            <i
              class="sc-kEqXSa bbXQBg"
            >
              <svg
                class="bi bi-chevron-down"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </i>
          </div>
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            3
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 3
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 3
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 3
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 3
          </div>
        </div>
        <div
          style="top: 475px; width: 635px; position: absolute; line-height: 35px; box-sizing: border-box; border-bottom: 1px solid #ddd; padding: 10px; height: 300px;"
        >
           
          这是一个展开的内容信息 
          {"height":35,"key":"key-3","cells":[{"name":"id","value":"3"},{"name":"userName","value":"my name 3"},{"name":"address","value":"address 3"},{"name":"email","value":"E-Mail 3"},{"name":"mobilePhone","value":"Mobile Phone 3"}],"object":{"id":"3","userName":"my name 3","address":"address 3","email":"E-Mail 3","mobilePhone":"Mobile Phone 3"}}
           
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 775px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            <i
              class="sc-kEqXSa bbXQBg"
            >
              <svg
                class="bi bi-chevron-right"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </i>
          </div>
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            4
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 4
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 4
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 4
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 4
          </div>
        </div>
        <div
          class="sc-hKFxyN fpSUVZ"
          role="row"
          style="height: 35px; top: 810px; width: 635px; line-height: 35px;"
        >
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 0px; z-index: 2; width: 35px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            <i
              class="sc-kEqXSa bbXQBg"
            >
              <svg
                class="bi bi-chevron-right"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </i>
          </div>
          <div
            class="sc-dlnjwi eDgJvg"
            role="gridcell"
            style="left: 35px; z-index: 2; width: 120px; height: 35px; line-height: 35px; position: sticky;"
            tabindex="-2"
          >
            5
          </div>
          <div
            class="sc-dlnjwi KaoYI"
            role="gridcell"
            style="left: 155px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            my name 5
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 275px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            address 5
          </div>
          <div
            class="sc-dlnjwi iGhXOb"
            role="gridcell"
            style="left: 395px; width: 120px; height: 35px; line-height: 35px;"
            tabindex="-2"
          >
            E-Mail 5
          </div>
          <div
            class="sc-dlnjwi sTOYU"
            role="gridcell"
            style="z-index: 2; width: 120px; height: 35px; line-height: 35px; float: right; right: 0px; position: sticky;"
            tabindex="-2"
          >
            Mobile Phone 5
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
